src/raftstore/store/peer.rs

```diff
@@ -686,30 +712,104 @@ impl Peer {
             }
             vec![]
         } else {
-            self.handle_raft_committed_entries(ready.committed_entries.take().unwrap())
+            let committed_entries = ready.committed_entries.take().unwrap();
+            // leader needs to update lease.
+            let mut to_be_updated = self.is_leader();
+            if !to_be_updated {
+                // It's not leader anymore, we are safe to clear proposals. If it becomes leader
+                // again, the lease should be updated when election is finished, old proposals
+                // have no effect. Although this may allow old uuids to be reused before old
+                // callbacks are called.
+                self.proposals.clear();
+            }
+            for entry in committed_entries.iter().rev() {
+                // raft meta is very small, can be ignored.
+                self.raft_log_size_hint += entry.get_data().len() as u64;
+                to_be_updated = to_be_updated &&
+                                self.maybe_update_lease(entry.get_term(), entry.get_data());
+            }
+            if !committed_entries.is_empty() {
+                self.handle_raft_committed_entries(committed_entries)
+            } else {
+                vec![]
+            }
         };
 
         self.raft_group.advance(ready);
     }
 
+    /// Try to update lease.
+    ///
+    /// If the it can make sure that its lease is the latest lease, returns true.
+    fn maybe_update_lease(&mut self, term: u64, data: &[u8]) -> bool {
+        let mut req = RaftCmdRequest::new();
+        let propose_time = match req.merge_from_bytes(data)
```

Comment:
Normally only the last entry will be decoded twice. The overhead can be omitted. Use term + index can find the propose time, but I still need to store uuid of every proposal already, so here just use uuid.